File: desktop/CredentialTyper.Transport/Crypto/CipherState.cs
using System.Buffers.Binary;
using System.Security.Cryptography;

namespace CredentialTyper.Transport.Crypto;

/// <summary>
/// Noise 的 CipherState：ChaCha20-Poly1305 + 单调递增 nonce。
/// Noise 的 96 位 nonce = 前 4 字节 0 || 小端 uint64 计数器。
/// </summary>
public sealed class CipherState
{
    private byte[]? _k;   // 32 字节；null = 无密钥（明文透传）
    private ulong _n;

    public void InitializeKey(byte[]? key)
    {
        _k = key;
        _n = 0;
    }

    public bool HasKey => _k is not null;

    public byte[] EncryptWithAd(ReadOnlySpan<byte> ad, ReadOnlySpan<byte> plaintext)
    {
        if (_k is null) return plaintext.ToArray();

        // 纯托管 RFC 8439 实现：不依赖平台 CNG，任何 Windows 版本都能与手机端互通。
        var ct = new byte[plaintext.Length];
        var tag = new byte[16];
        ManagedChaCha20Poly1305.Encrypt(_k, Nonce(_n), plaintext, ct, tag, ad);
        _n++;

        var outBuf = new byte[ct.Length + 16];
        ct.CopyTo(outBuf, 0);
        tag.CopyTo(outBuf, ct.Length);
        return outBuf;
    }

    public byte[] DecryptWithAd(ReadOnlySpan<byte> ad, ReadOnlySpan<byte> ciphertext)
    {
        if (_k is null) return ciphertext.ToArray();
        if (ciphertext.Length < 16) throw new CryptographicException("密文短于认证标签长度");

        int ctLen = ciphertext.Length - 16;
        var ct = ciphertext[..ctLen];
        var tag = ciphertext[ctLen..];
        var pt = new byte[ctLen];
        ManagedChaCha20Poly1305.Decrypt(_k, Nonce(_n), ct, tag, pt, ad);   // 标签不符会抛 CryptographicException
        _n++;
        return pt;
    }

    private static byte[] Nonce(ulong n)
    {
        var nonce = new byte[12];               // 前 4 字节保持 0
        BinaryPrimitives.WriteUInt64LittleEndian(nonce.AsSpan(4), n);
        return nonce;
    }
}

File: app/lib/noise/cipher_state.dart
import 'dart:typed_data';
import 'package:cryptography/cryptography.dart';

/// Noise 的 CipherState：ChaCha20-Poly1305 + 单调递增 nonce。
/// Noise 的 96 位 nonce = 前 4 字节 0 || 小端 uint64 计数器。
/// 对应桌面 `CipherState.cs`，密文布局 = ct || 16 字节 tag。
class CipherState {
  static final Chacha20 _aead = Chacha20.poly1305Aead();

  Uint8List? _k; // 32 字节；null = 无密钥（明文透传）
  int _n = 0;

  void initializeKey(Uint8List? key) {
    _k = key;
    _n = 0;
  }

  bool get hasKey => _k != null;

  Future<Uint8List> encryptWithAd(List<int> ad, List<int> plaintext) async {
    if (_k == null) return Uint8List.fromList(plaintext);

    final box = await _aead.encrypt(
      plaintext,
      secretKey: SecretKey(_k!),
      nonce: _nonce(_n),
      aad: ad,
    );
    _n++;

    final out = Uint8List(box.cipherText.length + 16);
    out.setAll(0, box.cipherText);
    out.setAll(box.cipherText.length, box.mac.bytes);
    return out;
  }

  Future<Uint8List> decryptWithAd(List<int> ad, Uint8List ciphertext) async {
    if (_k == null) return Uint8List.fromList(ciphertext);
    if (ciphertext.length < 16) {
      throw const FormatException('密文短于认证标签长度');
    }

    final ctLen = ciphertext.length - 16;
    final box = SecretBox(
      ciphertext.sublist(0, ctLen),
      nonce: _nonce(_n),
      mac: Mac(ciphertext.sublist(ctLen)),
    );
    final pt = await _aead.decrypt(box, secretKey: SecretKey(_k!), aad: ad);
    _n++;
    return Uint8List.fromList(pt);
  }

  static Uint8List _nonce(int n) {
    final nonce = Uint8List(12); // 前 4 字节保持 0
    ByteData.sublistView(nonce).setUint64(4, n, Endian.little);
    return nonce;
  }
}
